src/interp.rs

```diff
@@ -34,36 +34,89 @@
 // OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN THE
 // SOFTWARE.
 
+use byteorder::{ByteOrder, LittleEndian};
 use std::process::exit;
 use rustc::ty;
 use rustc::ty::{Ty, TyCtxt, TypeVariants};
 use rustc::hir::def_id::DefId;
+use rustc::middle::const_val::ConstVal;
 use rustc::mir::{
     START_BLOCK, Mir, BasicBlock, BasicBlockData, Statement, StatementKind,
     Rvalue, Place, Terminator, TerminatorKind, AggregateKind, Operand, Constant,
-    Literal
+    Literal, BinOp
 };
+use syntax::ast::IntTy;
 
 use machine;
 use machine::{Machine, Address};
-use primitive::ByteCast;
 
 const MEMORY_CAPACITY: usize = 1024; // in Bytes.
 
-/// Stores a value represented as a vector of bytes, along with its type
-/// annotation. It's important to keep track of a value's type while it is
-/// being evaluated, so that we know how to interpret the bytes.
 #[derive(Debug, Clone)]
-pub struct TypedVal<'tcx> {
-    pub ty: Ty<'tcx>,
-    pub val: Vec<u8>,
+pub enum Value {
+    Int(u128),
+    Ref(usize),
+    None,
 }
 
+#[derive(Debug, Clone)]
+pub struct TyVal<'tcx> {
+    ty: Ty<'tcx>,
+    val: Value,
+}
+
+impl<'tcx> TyVal<'tcx> {
+    pub fn new(ty: Ty<'tcx>, val: Value) -> Self {
+        Self {
+            ty: ty,
+            val: val
+        }
+    }
+}
+
+impl<'tcx> TyVal<'tcx> {
+    pub fn to_bytes(&self) -> Vec<u8> {
+        match self.ty.sty {
+            TypeVariants::TyInt(int_ty) => {
+                let val = match self.val {
+                    Value::Int(i) => i,
+                    _ => panic!("Mismatched Types")
+                };
+
+                match int_ty {
+                    IntTy::I32 => {
+                        let mut buf = [0;4];
+                        LittleEndian::write_i32(&mut buf, val as i32);
+                        buf.to_vec()
```

Comment:
So this is the return value here? It's quite hard to see because it's so deeply nested. Not sure if there's much you could do about that other than use an early return.

Reply:
Yeah, I agree it's hard to see. I can use the return statement for clarity.

Reply:
Ask Laurie. He's a expression purist :P

Reply:
I've pushed a fix including an early return statement. I think it reads better. @ltratt Do you have any thoughts on this?